packages/manager/src/features/Lish/Weblish.tsx

```diff
@@ -56,24 +62,75 @@ export class Weblish extends React.Component<Props, State> {
   connect() {
     const { weblish_url, ws_protocols } = this.props;
 
-    this.socket = new WebSocket(weblish_url, ws_protocols);
+    /* When this.socket != origSocket, the socket from this connect()
+     * call has been closed and possibly replaced by a new socket. */
+    const origSocket = new WebSocket(weblish_url, ws_protocols);
+    this.socket = origSocket;
+
+    this.lastMessage = '';
+    this.setState({error: ''});
 
     this.socket.addEventListener('open', () => {
       if (!this.mounted) {
         return;
       }
-      this.setState({ renderingLish: true }, () => this.renderTerminal());
+      this.setState({ renderingLish: true }, () => this.renderTerminal(origSocket));
+    });
+
+    this.socket.addEventListener('close', (evt) => {
+      if (this.socket !== origSocket) {
+        return;
+      }
+      this.socket = null;
+      this.terminal?.dispose();
+      this.setState({ renderingLish: false });
+      if (!this.mounted) {
+        return;
+      }
+
+      let parsed = null;
+      try {
+        if (evt?.reason) {
+          parsed = JSON.parse(evt.reason);
+        }
+        if (parsed === null) {
+          parsed = JSON.parse(this.lastMessage);
+        }
+      } catch {
+      }
+
+      if (this.retryLimiter.retryAllowed()) {
+        if (parsed?.errors?.[0]?.reason === "expired" ||
+            (parsed?.type === "error" &&
+             typeof parsed?.reason === "string" &&
+             parsed?.reason.toLowerCase() === "your session has expired.")) {
+          const { refreshToken } = this.props;
+          refreshToken();
+        } else {
+          this.connect();
+        }
+      } else {
+        this.setState({error: formatError(parsed, "Unexpected WebSocket close")});
+      }
```

Comment:
Can we unnest this a bit?
```suggestion
      if (!this.retryLimiter.retryAllowed()) {
        return this.setState({
          error: formatError(parsed, 'Unexpected WebSocket close'),
        });
      }

      if (
        parsed?.errors?.[0]?.reason === 'expired' ||
        (parsed?.type === 'error' &&
          typeof parsed?.reason === 'string' &&
          parsed?.reason.toLowerCase() === 'your session has expired.')
      ) {
        const { refreshToken } = this.props;
        refreshToken();
      } else {
        this.connect();
      }
```

Reply:
Flattened. I also refactored the error parsing/formatting some, so hopefully everything is much cleaner & clearer now!